File: go/src/RumbleRacing/core/asset/o3d/geometry.go
package o3d

import "fmt"

type Vertex struct {
	X float32
	Y float32
	Z float32
}

type Normal struct {
	ADCBitSet bool

	X float32
	Y float32
	Z float32
}

type UV struct {
	U float32
	V float32
}

type Geometry struct {
	Buffers []Buffer
}

type PrimitiveType uint8

func (t *PrimitiveType) String() string {
	switch *t {
	case Triangle:
		return "Triangle"
	case TriangleStrip:
		return "TriangleStrip"
	default:
		panic("Unhandled type")
	}
}

const (
	Triangle      PrimitiveType = 0b011
	TriangleStrip PrimitiveType = 0b100
)

type Primitive struct {
	TotalVertsInPrimitive int
	PrimType              PrimitiveType

	Vertices []Vertex
	Normals  []Normal
	UVs      []UV
}

type Buffer struct {
	NumHeaderLines int
	NumStrips      int
	Primitives     []Primitive

	TextureId int // -1 if not found
}

type bufferChunks struct {
	BufferChunks []bufferChunk
}
type bufferChunk struct {
	BufferHeader []V4_32Entry
	Strips       []stripChunk
}

type triple struct {
	A UnpackData
	B UnpackData
	C UnpackData
}

type stripChunk struct {
	GIFTag V4_32Entry

	// DataTriples stores groups of (A, B, C) unpacks belonging to this header
	DataTriples []triple
}

func GetGeometry(commandStream []VifCommand, textures TextureMeta) (*Geometry, error) {
	var filtered []VifCommand
	for _, cmd := range commandStream {
		if cmd.Kind == VifCommandUNPACK {
			filtered = append(filtered, cmd)
		}
	}

	chunks, err := getBufferChunks(filtered)
	if err != nil {
		return nil, err
	}

	geometry := &Geometry{}

	for _, bChunk := range chunks.BufferChunks {
		buf := Buffer{
			NumStrips:      int(bChunk.BufferHeader[0].V1),
			NumHeaderLines: int(bChunk.BufferHeader[0].V2),
			TextureId:      -1,
		}

		buf.TextureId = findTextureId(textures, bChunk.BufferHeader[len(bChunk.BufferHeader)-1].Offset)

		for _, sChunk := range bChunk.Strips {
			strip := Primitive{
				TotalVertsInPrimitive: int(sChunk.GIFTag.V1 & 0x7fff),
				PrimType:              PrimitiveType(uint8((sChunk.GIFTag.V2 & (0b111 << 15) >> 15))),
			}

			// Process every triple associated with this strip header
			for _, triple := range sChunk.DataTriples {
				// if len(triple) < 3 {
				// 	continue // Or handle unexpected partial data
				// }

				cmdA, cmdB, cmdC := triple.A, triple.B, triple.C

				switch {
				case cmdA.Type == UnpackTypeV3_32 && cmdB.Type == UnpackTypeV3_32 && cmdC.Type == UnpackTypeV2_32:
					for j := 0; j < len(cmdA.V3_32); j++ {
						strip.Normals = append(strip.Normals, Normal{X: cmdA.V3_32[j].V1, Y: cmdA.V3_32[j].V2, Z: cmdA.V3_32[j].V3, ADCBitSet: cmdA.V3_32[j].ADCBitSet})
						strip.Vertices = append(strip.Vertices, Vertex{X: cmdB.V3_32[j].V1, Y: cmdB.V3_32[j].V2, Z: cmdB.V3_32[j].V3})
						strip.UVs = append(strip.UVs, UV{U: cmdC.V2_32[j].V1, V: cmdC.V2_32[j].V2})
					}

				case cmdA.Type == UnpackTypeV3_32 && cmdB.Type == UnpackTypeV2_32 && cmdC.Type == UnpackTypeV4_8:
					for j := 0; j < len(cmdA.V3_32); j++ {
						strip.Vertices = append(strip.Vertices, Vertex{X: cmdA.V3_32[j].V1, Y: cmdA.V3_32[j].V2, Z: cmdA.V3_32[j].V3})
						strip.UVs = append(strip.UVs, UV{U: cmdB.V2_32[j].V1, V: cmdB.V2_32[j].V2})
						strip.Normals = append(strip.Normals, Normal{
							X:         float32(cmdC.V4_8[j].V1) / 255.0,
							Y:         float32(cmdC.V4_8[j].V2) / 255.0,
							Z:         float32(cmdC.V4_8[j].V3) / 255.0,
							ADCBitSet: cmdC.V4_8[j].ADCBitSet,
						})
					}
				}
			}
			buf.Primitives = append(buf.Primitives, strip)
		}
		geometry.Buffers = append(geometry.Buffers, buf)
	}

	return geometry, nil
}

func getBufferChunks(filtered []VifCommand) (*bufferChunks, error) {
	result := &bufferChunks{}
	i := 0

	for i < len(filtered) {
		// Buffer Boundary: Two consecutive V4_32s
		if i+1 < len(filtered) &&
			filtered[i].Unpack.Type == UnpackTypeV4_32 &&
			filtered[i+1].Unpack.Type == UnpackTypeV4_32 {

			chunk := bufferChunk{
				BufferHeader: filtered[i].Unpack.V4_32,
			}
			i++ // Advance past the buffer header

			// Collect strips within this buffer
			for i < len(filtered) {
				// Stop if we hit the next buffer (two V4_32s)
				if i+1 < len(filtered) &&
					filtered[i].Unpack.Type == UnpackTypeV4_32 &&
					filtered[i+1].Unpack.Type == UnpackTypeV4_32 {
					break
				}

				// Expect a Strip Header
				if filtered[i].Unpack.Type != UnpackTypeV4_32 {
					return nil, fmt.Errorf("expected strip header at index %d", i)
				}

				sChunk := stripChunk{}
				if len(filtered[i].Unpack.V4_32) > 0 {
					sChunk.GIFTag = filtered[i].Unpack.V4_32[0]
				}
				i++

				// Collect all following A, B, C triples until the next V4_32 header
				for i < len(filtered) && filtered[i].Unpack.Type != UnpackTypeV4_32 {
					tripl := make([]UnpackData, 0, 3)
					for j := 0; j < 3 && i < len(filtered); j++ {
						if filtered[i].Unpack.Type == UnpackTypeV4_32 {
							break
						}
						tripl = append(tripl, *filtered[i].Unpack)
						i++
					}
					if len(tripl) > 2 {
						sChunk.DataTriples = append(sChunk.DataTriples, triple{A: tripl[0], B: tripl[1], C: tripl[2]})
					} else {
						// TODO: figure out why this happens
						// fmt.Println("Missing triple!")
					}
				}
				chunk.Strips = append(chunk.Strips, sChunk)
			}
			result.BufferChunks = append(result.BufferChunks, chunk)
		} else {
			i++
		}
	}
	return result, nil
}

func findTextureId(textures TextureMeta, dataAddress uint64) int {
	bestId := -1
	var bestOffset uint64

	for _, entry := range textures.TextureEnties {
		if uint64(entry.ELDAOffset) <= dataAddress {
			if bestId == -1 || uint64(entry.ELDAOffset) > bestOffset {
				bestOffset = uint64(entry.ELDAOffset)
				bestId = entry.TextureId
			}
		}
	}
	return bestId
}

File: src/RumbleRacing/asset/o3d/geometry.ts
import { vec2, vec3 } from "gl-matrix";
import { VifCmd, VifUnpackFormat } from "../../../Common/PS2/VIF";
import { VifCommand, UnpackData } from "./vif";

export interface Primitive {
  totalVertsInPrimitive: number;
  primType: number;
  vertices: {
    position: vec3;
    normal: vec3;
    uv: vec2;
    adcBitSet: boolean;
  }[];
}

export interface Buffer {
  numHeaderLines: number;
  numStrips: number;
  primitives: Primitive[];
  textureId: number;
}

export interface Geometry {
  buffers: Buffer[];
}

interface Triple {
  a: UnpackData;
  b: UnpackData;
  c: UnpackData;
}

interface StripChunk {
  gifTagV1: number;
  gifTagV2: number;
  dataTriples: Triple[];
}

interface BufferChunk {
  bufferHeader: { v1: number; v2: number; lastOffset: number }[];
  strips: StripChunk[];
}

export interface TextureEntry {
  eldaOffset: number;
  textureId: number;
}

export interface TextureMeta {
  numTextures: number;
  textureEntries: TextureEntry[];
}

function getBufferChunks(filtered: VifCommand[]): BufferChunk[] {
  const result: BufferChunk[] = [];
  let i = 0;

  while (i < filtered.length) {
    if (
      i + 1 < filtered.length &&
      filtered[i].unpack?.type === VifUnpackFormat.V4_32 &&
      filtered[i + 1].unpack?.type === VifUnpackFormat.V4_32
    ) {
      const bufHeader = filtered[i].unpack!.v4_32.map((e) => ({
        v1: e.v1,
        v2: e.v2,
        lastOffset: e.offset,
      }));
      i++;

      const strips: StripChunk[] = [];

      while (i < filtered.length) {
        if (
          i + 1 < filtered.length &&
          filtered[i].unpack?.type === VifUnpackFormat.V4_32 &&
          filtered[i + 1].unpack?.type === VifUnpackFormat.V4_32
        ) {
          break;
        }

        if (filtered[i].unpack?.type !== VifUnpackFormat.V4_32) {
          throw new Error(`expected strip header at index ${i}`);
        }

        const gifEntries = filtered[i].unpack!.v4_32;
        const sChunk: StripChunk = {
          gifTagV1: gifEntries.length > 0 ? gifEntries[0].v1 : 0,
          gifTagV2: gifEntries.length > 0 ? gifEntries[0].v2 : 0,
          dataTriples: [],
        };
        i++;

        while (
          i < filtered.length &&
          filtered[i].unpack?.type !== VifUnpackFormat.V4_32
        ) {
          const tripl: UnpackData[] = [];
          for (let j = 0; j < 3 && i < filtered.length; j++) {
            if (filtered[i].unpack?.type === VifUnpackFormat.V4_32) break;
            tripl.push(filtered[i].unpack!);
            i++;
          }
          if (tripl.length > 2) {
            sChunk.dataTriples.push({ a: tripl[0], b: tripl[1], c: tripl[2] });
          }
        }
        strips.push(sChunk);
      }

      const lastHeader = bufHeader[bufHeader.length - 1];
      result.push({ bufferHeader: bufHeader, strips });
    } else {
      i++;
    }
  }

  return result;
}

function findTextureId(textures: TextureMeta, dataAddress: number): number {
  let bestId = -1;
  let bestOffset = -1;

  for (const entry of textures.textureEntries) {
    if (entry.eldaOffset <= dataAddress) {
      if (bestId === -1 || entry.eldaOffset > bestOffset) {
        bestOffset = entry.eldaOffset;
        bestId = entry.textureId;
      }
    }
  }

  return bestId;
}

export function getGeometry(
  commandStream: VifCommand[],
  textures: TextureMeta,
): Geometry {
  const filtered = commandStream.filter((c) => c.kind === VifCmd.UNPACK_MASK);
  const chunks = getBufferChunks(filtered);
  const geometry: Geometry = { buffers: [] };

  for (const bChunk of chunks) {
    const lastHeader = bChunk.bufferHeader[bChunk.bufferHeader.length - 1];
    const buf: Buffer = {
      numStrips: bChunk.bufferHeader[0]?.v1 ?? 0,
      numHeaderLines: bChunk.bufferHeader[0]?.v2 ?? 0,
      primitives: [],
      textureId: findTextureId(textures, lastHeader?.lastOffset ?? 0),
    };

    for (const sChunk of bChunk.strips) {
      const strip: Primitive = {
        totalVertsInPrimitive: sChunk.gifTagV1 & 0x7fff,
        primType: (sChunk.gifTagV2 & (0b111 << 15)) >> 15,
        vertices: [],
      };

      for (const triple of sChunk.dataTriples) {
        const { a: cmdA, b: cmdB, c: cmdC } = triple;

        if (
          cmdA.type === VifUnpackFormat.V3_32 && // A = normals
          cmdB.type === VifUnpackFormat.V3_32 && // B = vertices
          cmdC.type === VifUnpackFormat.V2_32 // C = uvs
        ) {
          for (let j = 0; j < cmdA.v3_32.length; j++) {
            strip.vertices.push({
              normal: vec3.fromValues(
                cmdA.v3_32[j].v1,
                cmdA.v3_32[j].v2,
                cmdA.v3_32[j].v3,
              ),
              adcBitSet: cmdA.v3_32[j].adcBitSet,
              position: vec3.fromValues(
                cmdB.v3_32[j].v1,
                cmdB.v3_32[j].v2,
                cmdB.v3_32[j].v3,
              ),
              uv: vec2.fromValues(cmdC.v2_32[j].v1, cmdC.v2_32[j].v2),
            });
          }
        } else if (
          cmdA.type === VifUnpackFormat.V3_32 && // A = vertices
          cmdB.type === VifUnpackFormat.V2_32 && // B = uvs
          cmdC.type === VifUnpackFormat.V4_8 // C = normals
        ) {
          for (let j = 0; j < cmdA.v3_32.length; j++) {
            strip.vertices.push({
              position: vec3.fromValues(
                cmdA.v3_32[j].v1,
                cmdA.v3_32[j].v2,
                cmdA.v3_32[j].v3,
              ),
              uv: vec2.fromValues(cmdB.v2_32[j].v1, cmdB.v2_32[j].v2),
              normal: vec3.fromValues(
                cmdC.v4_8[j].v1 / 255.0,
                cmdC.v4_8[j].v2 / 255.0,
                cmdC.v4_8[j].v3 / 255.0,
              ),
              adcBitSet: cmdC.v4_8[j].adcBitSet,
            });
          }
        }
      }

      buf.primitives.push(strip);
    }

    geometry.buffers.push(buf);
  }

  return geometry;
}
